core/os/android/adb/screen.go

```diff
@@ -17,71 +17,94 @@ package adb
 import (
 	"context"
 	"regexp"
+	"time"
 
 	"github.com/google/gapid/core/fault"
 	"github.com/google/gapid/core/log"
 	"github.com/google/gapid/core/os/android"
 )
 
 const (
-	ErrScreenOnState   = fault.Const("Couldn't get screen-on state")
-	ErrLockScreenState = fault.Const("Couldn't get lockscreen state")
+	screenOnLocked = iota
+	screenOnUnlocked
+	screenOffLocked
+	screenOffUnlocked
 )
 
-var displayOnRegex = regexp.MustCompile("Display Power: state=(ON|DOZE|OFF)")
-var displayReadyRegex = regexp.MustCompile("mDisplayReady=(true|false)")
+const (
+	keyEventEffectDelay time.Duration = 500 * time.Millisecond
+)
+
+const (
+	ErrScreenState = fault.Const("Couldn't get screen state")
+)
+
+var screenStateRegex = regexp.MustCompile("mScreenState=(ON|OFF)_(LOCKED|UNLOCKED)")
 
-// IsScreenOn returns true if the device's screen is currently on.
-func (b *binding) IsScreenOn(ctx context.Context) (bool, error) {
-	res, err := b.Shell("dumpsys", "power").Call(ctx)
+func (b *binding) getScreenState(ctx context.Context) (int, error) {
+	res, err := b.Shell("dumpsys", "nfc").Call(ctx)
 	if err != nil {
-		return false, err
+		return -1, err
 	}
-	switch displayOnRegex.FindString(res) {
-	case "Display Power: state=ON":
-		switch displayReadyRegex.FindString(res) {
-		case "mDisplayReady=true":
-			return true, nil
-		case "mDisplayReady=false":
-			return false, nil
-		}
-	case "Display Power: state=DOZE":
-		return false, nil
-	case "Display Power: state=OFF":
-		return false, nil
+	switch screenStateRegex.FindString(res) {
+	case "mScreenState=ON_LOCKED":
+		return screenOnLocked, nil
+	case "mScreenState=ON_UNLOCKED":
+		return screenOnUnlocked, nil
+	case "mScreenState=OFF_LOCKED":
+		return screenOffLocked, nil
+	case "mScreenState=OFF_UNLOCKED":
+		return screenOffUnlocked, nil
 	}
-	return false, log.Err(ctx, ErrScreenOnState, "")
+	return -1, log.Err(ctx, ErrScreenState, "")
 }
 
-// TurnScreenOn turns the device's screen on.
-func (b *binding) TurnScreenOn(ctx context.Context) error {
-	if isOn, err := b.IsScreenOn(ctx); err != nil || isOn {
-		return err
+// IsScreenUnlocked returns true is the device's screen is on and unlocked.
+func (b *binding) IsScreenUnlocked(ctx context.Context) (bool, error) {
+	screenState, err := b.getScreenState(ctx)
+	if err != nil {
+		return false, err
 	}
-	return b.KeyEvent(ctx, android.KeyCode_Power)
+	return screenState == screenOnUnlocked, nil
 }
 
-// TurnScreenOff turns the device's screen off.
-func (b *binding) TurnScreenOff(ctx context.Context) error {
-	if isOn, err := b.IsScreenOn(ctx); err != nil || !isOn {
-		return err
-	}
-	return b.KeyEvent(ctx, android.KeyCode_Power)
-}
+// UnlockScreen tries to make the device's screen to be on and unlocked.
+func (b *binding) UnlockScreen(ctx context.Context) error {
```

Comment:
I think it makes sense to return an error when the screen is not in the `screenOnUnlocked` state after this call. That way we don't have to follow the `UnlockScreen`  then check `IsScreenUnlocked` pattern. If we want callers to be able to distinguish between the "i tried, it didn't work" error and other errors, you can make the returned error a public constants (like, say, `io.EOF` et al)

Reply:
Good point, I adapted this way: UnlockScreen() now returns (bool, error) to indicate whether is successfully managed to unlock the screen.